src/lib/reasoners/bitv.ml

```diff
@@ -784,36 +802,38 @@ module Shostak(X : ALIEN) = struct
             let c, eq, pat = apply_sub_rev bsub c [] [] req in
             eq, csub, if c then Some pat else None
         end
-      | st :: eq, n :: pt when st.sz < n ->
+      | { bv = S_Var { sorte = B; _ }; _ } as st :: eq,
+        n :: pt when st.sz < n ->
```

Comment:
Why can't we use `slice_var` here?

Reply:
Also why doesn't `slice_var` return the `changed` boolean instead of having a special case for B-vars?

Reply:
> Why can't we use slice_var here?

Because we're not slicing variables, this is the result of a previous slicing and substitution of a B variable.

> Also why doesn't slice_var return the changed boolean instead of having a special case for B-vars?

It does make sens to make `slice_var` return `changed`, but since `apply_sub_rev` also returns `changed` anyway, though it wouldn't be necessary, since `apply_sub_rev` would do it anyway.
The special case for B is due to the fact that that case can only be reached with B variable that was the result of a slicing and a substitution. (This also returns `changed = true` which is weird, correct but probably unnecessary, I think it would be "prettier" to have only `apply_sub_rev` return `changed` that way the argument can be removed since `apply_sub_rev` is only applied at the end)

Reply:
Ah right, I got confused by the change that added the `sorte = B` case, thanks for the clarification. We should either perform the splitting (as was done before) or have an `assert false` here if the we get `st.sz < n` and the `sorte` is not `B` rather than fall through to the slicing case below.

Regarding returning `changed = true` here — you are right it is probably unnecessary because the `bsub` must be non-empty if we reach that case (when I wrote the code, I initially planned on removing `slicing_pattern`, in which case `changed = true` here would have been necessary). I am OK with not setting `changed` here but we probably should assert that `bsub` is nonempty in that case, to ensure that `apply_sub_rev` actually gets called.

Reply:
Right, It's true that the slicing case doesn't have a `when st.sz > n`, so yeah, an assertion is better. Done